doc/design/draft/named-segments.md

```diff
@@ -0,0 +1,228 @@
+name instead of by access control. This will increase flexibility while still granting all the benefits of the current access control model.
+
+## Requirements
+
+* The RouDi configuration must support specifying names for shared memory segments.
+* If a name is not configured, the fallback behavior must be the same as it is currently - assigning the name of the current process user.
+* Communication endpoints should be able to specify which segments they intend to use.
+  * Publishers may specify a single segment into which they will publish.
+  * Subscribers may specify one or more segments from which they may receive messages.
+  * Clients and Servers may specify a single segment into which they send requests/responses, and multiple segments from which they receive responses/requests.
+* When no segment is specified, the fallback behavior must be the same as it is currently
+  * Publishers will select the *unique* segment they have write access to.
+  * Subscribers will receive messages from all segments they have read access to.
+  * Clients and Servers follow the same rules as Publishers for data they send and the same rules as Subscribers for data they receive.
+* Creation will fail if a communication endpoint requests to use a segment to which it does not
+have the proper access rights.
+
+### Optional additional requirements
+
+* When initializing the runtime, a list of segment names may be provided. In this case, rather than mapping all shared memory segments available in the system, only those that are named shall be mapped.
+  * When a communication endpoint requests a segment that has not been mapped, this may be configured to either result in a failure or in mapping the requested segment on the fly.
+
+## Design
+
+### Updated RouDi Config
+
+Supporting named segments will require add an additional `name` field to the RouDi config under the `[[segment]]` heading, as well as updating the config version Example:
+
+```
+[general]
+version = 2
+
+[[segment]]
+name = "foo"
+
+[[segment.mempool]]
+size = 32
+count = 10000
+```
+
+Updating the version will be necessary for older versions of RouDi to fail informatively when presented with a newer config. Otherwise RouDi can support version 1 configs as if they are version 2 configs with no `name` fields specified.
+
+### Requesting Segments
+
+#### When creating a Publisher
+
+A new field will be added to the [PublisherOptions struct](../../../iceoryx_posh/include/iceoryx_posh/popo/publisher_options.hpp) as follows:
+
+```
+struct PublisherOptions
+{
+  ...
+  ShmName_t segmentName{""};
+```
+
+#### When creating a Subscriber
+
+A similar field will be added to the [SubscriberOptions struct](../../../iceoryx_posh/include/iceoryx_posh/popo/subscriber_options.hpp), except that it will support multiple elements
+
+```
+struct SubscriberOptions
+{
+  ...
+  vector<ShmName_t, MAX_SUBSCRIBER_SEGMENTS> segmentNames{};
```

Comment:
Here we have again the issue with the fixed size container. We need to find a sane limit without causing the data structures exploding in their memory requirements.

Reply:
Yes... 🙁 

In my own experimentation I actually just made it a single string. I think in practice that is usually what would be used since the main benefit of this feature is communicating over specific isolated "channels". But I was trying to leave a bit flexibility for some hypothetical use cases like having multiple publishers publishing on different segments, but the same topic. 

Another idea to consider which I do not recall whether I've mentioned before, but we could instead of introducing a new string field, add some coupling to the service description. For example - we could say that the "group" field is used to identify the shared memory segment to use, and all topics belonging to the same group share the same segment. 

I'm very hesitant/skeptical of this approach because I prefer not to couple features to one another. This would also more likely be a breaking API change unless we were careful to add some flags like "useServiceGroupAsSegmentMapping" or whatever.

Reply:
I guess you could have a set of flags / bitmap that with mutual exclusivity allows to choose from the following:
```
noSegmentMapping
groupAsSegmentMapping
instanceAsSegmentMapping
topicAsSegmentMapping
```
(topic is fully qualified on the group and instance as well)

Reply:
I think, in terms of the FFI implications here. In a system where multiple producers for a single topic is allowed, you would still want there to be a way for each producer to be isolated to writing to different shared memory segments and as such the "match by service description" approach would be insufficient.

Reply:
I'm wondering whether it is necessary to have the segments in the `SubscriberOptions` at all. An alternative would be to leave it to the runtime, e.g.
```cpp
for(const auto* name: {"foo", "bar"}) {
    runtime.mapReadOnlySegment(name);
}
```
The same could be done for the publisher. Since the whole process has read/write access to these segments and there might be multiple subscriber/publisher which access the same segments this could be done upfront at the start of the application.

Regarding the `ServiceDescription`. There is the idea to combine the three string into one. See also https://github.com/eclipse-iceoryx/iceoryx/issues/1052. The three separate strings date back to a time where iceoryx was meant to be the base for an AUTOSAR Adaptiv middleware and some design decisions where made with that idea in mind. With iceoryx2 there is also just one string and with a new API we could also move into this direction. The old `ServiceDescription` would the just be concatenated to e.g. `/service/event/instance`.

For iceoryx2 we currently have a separate shm segment for each publisher. Since this can become an issue with many publisher, there is the idea to group some together and create a `SharedMemoryDirectory` which can be used by multiple publisher. I think this goes into the same direction of what you describe as "group". I need to check with @elfenpiff how exactly this is implemented with iceoryx2.

I think we can hit two birds with this refactoring and create a API to ease the transition from iceoryx1 to iceoryx2.

Reply:
Yeah so let me think out loud a little here.

- If we do not specify which segments we expect the subscriber to receive messages from, it could receive them from any of the mapped segments.
- If we then connect to a channel whose segment has not been mapped, we need to either map it dynamically, or ensure some sort of error is triggered. This follows from the other comment you make about having an enum. 
- We will lose the ability somewhat to control who we may allow to talk to a subscriber. Though I suppose this can be solved through other means of controlling which channels use which segments (e.g. segment names are coupled to channels through some configuration scheme so you know ahead of time anyway that channel data can only come across a specific segment). 

I think dropping the explicit segment list from the subscriber works for my purposes and is simpler. The other somewhat ambiguous concerns I'm surmising here can be readdressed if necessary.

Reply:
The other tricky thing I've encountered is the fact that a subscriber may succceed in "connecting" with a publisher, but since the publisher publishes messages to a segment the subscriber has no access to, it doesn't actually receive any messages, which can be a really surprising silent error.

Reply:
For example, if we added a list of segment ids to the subscriber port data, and a single segment id to the publisher port data, we could use this additional information in [isCompatiblePubSub](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/source/roudi/port_manager.cpp#L521) to determine if a publishers messages can actually reach a subscriber.

And we can then treat this error the same way we treat policy mismatches.

Reply:
Also just thinking out loud. The question is whether this check should be done in RouDi or the subscriber. A subscriber might receive a sample from a segment which is not mapped but the process might have read access. The subscriber process would then print an error message if it could not map the segment. This would be more visible to the user than a warning from RouDi. We already had some bug reports because of RouDi printing an error message and the user just looking at the terminal with the application.

Reply:
So I'll need to take another look but IIRC one issue I ran into was that the subscriber didn't have direct access to that sort of information. The Runtime is what has access to a lot of the information and the only interaction Subscribers have with it is in the beginning via `getMiddlewareSubscriber()` to get the subscriber port. 

The subscriber could certainly tell that the segment is not mapped, but the only way to get to the information identifying the segment to determine if the subscriber in fact has read access is via the segment manager, which the runtime currently only temporarily gains access to during construction (the identifying information is collected via the IPC connection, but the runtime doesn't hold onto it).

I must confess that in following your suggestion to subclass things, I managed to subclass the runtime and, after moving the [ipcChannelInterface](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp#L114) from private to protected, have now stored the relative pointer to the segment manager as a member of my derived class. 

With this, you CAN easily ask questions like "Do I have read access to this segment?" And then decide "well, lets go map it now then!". But to do this - the subscriber needs to talk to the runtime. I suppose this is alright - it can always invoke the runtime via `getInstance()`. If we ever get to refactoring the runtime away from being a singleton we would instead need to reference it as a member of the subscriber, and maybe do some attorney pattern stuff for the API's touching the segment manager that only the subscriber should be using in only a specific way. 

As for the runtime - I'd recommend taking a look at this `shared_memory_user` and considering what we need to change to expose the info we need, without exposing too much. It seems to try to only temporarily access the segment manager to just get the info it needs and map everything, and then relinquish it back. But if we want to be able to map segments later - we are going to need to access the segment manager again anyway so... either the runtime should just keep a hold of the segment manager or the `shared_memory_user` API also needs to be extended. I guess there is no reason that class can't privately reference the segment manager past the life of the constructor 🤔

Reply:
Oh another thing I wanted to mention. Having access to the segment manager seems to make accessing introspection information so much easier than setting up a whole other introspection client. Haha.